File: leetcode/easy/SubtreeOfAnotherTree.py
from typing import Optional



# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        """
        Checks if a binary tree `subRoot` is a subtree of another binary tree `root`.

        A subtree of a tree `T` is a tree consisting of a node in `T` and all of its
        descendants in `T`.

        Algorithm:
        1. The main function `isSubtree` traverses the `root` tree recursively.
        2. For each node in `root`, it calls a helper function `_isSameTree` to check
           if the subtree starting at that node is identical to `subRoot`.
        3. If an identical subtree is found, we return True.
        4. If not, we continue the search by recursively calling `isSubtree` on the
           left and right children of the current node.
        5. The recursion stops when we run out of nodes in the `root` tree.

        Time Complexity: O(M * N), where M is the number of nodes in `root` and N is
        the number of nodes in `subRoot`. In the worst case, we call `_isSameTree` (O(N))
        for every node in `root` (O(M)).
        Space Complexity: O(H), where H is the height of the `root` tree, due to the
        recursion stack.
        """
        # Base case: An empty tree is a subtree of any tree.
        if not subRoot:
            return True
        # Base case: If the main tree is empty but subRoot is not, it's impossible.
        if not root:
            return False

        # Check if the tree starting at the current `root` node is the same as `subRoot`.
        if self._isSameTree(root, subRoot):
            return True

        # If not, recursively check if `subRoot` is a subtree of the left OR right child.
        return self.isSubtree(root.left, subRoot) or self.isSubtree(root.right, subRoot)

    def _isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        """Helper function to check if two trees are structurally and valuably identical."""
        # If both nodes are None, they are identical.
        if not p and not q:
            return True
        # If one node is None, or their values differ, they are not identical.
        if not p or not q or p.val != q.val:
            return False

        # Recursively check the left and right subtrees.
        return self._isSameTree(p.left, q.left) and self._isSameTree(p.right, q.right)

File: leetcode/easy/SubtreeOfAnotherTree.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    /**
     * Checks if a binary tree `subRoot` is a subtree of another binary tree `root`.
     *
     * A subtree of a tree `T` is a tree consisting of a node in `T` and all of its
     * descendants in `T`.
     *
     * Algorithm:
     * 1. The main function `isSubtree` traverses the `root` tree recursively.
     * 2. For each node in `root`, it calls a helper function `isSameTree` to check
     *    if the subtree starting at that node is identical to `subRoot`.
     * 3. If an identical subtree is found, we return true.
     * 4. If not, we continue the search by recursively calling `isSubtree` on the
     *    left and right children of the current node.
     * 5. The recursion stops when we run out of nodes in the `root` tree.
     *
     * @param root The root of the main binary tree.
     * @param subRoot The root of the potential subtree.
     * @return True if `subRoot` is a subtree of `root`, false otherwise.
     *
     * Time Complexity: O(M * N), where M is the number of nodes in `root` and N is
     *                  the number of nodes in `subRoot`. In the worst case, we call
     *                  `isSameTree` (O(N)) for every node in `root` (O(M)).
     * Space Complexity: O(H), where H is the height of the `root` tree, due to the
     *                   recursion stack.
     */
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        // Base case: An empty tree is a subtree of any tree.
        if (subRoot == null) {
            return true;
        }
        // Base case: If the main tree is empty but subRoot is not, it's impossible.
        if (root == null) {
            return false;
        }

        // Check if the tree starting at the current `root` node is the same as `subRoot`.
        if (isSameTree(root, subRoot)) {
            return true;
        }

        // If not, recursively check if `subRoot` is a subtree of the left OR right child.
        return isSubtree(root.left, subRoot) || isSubtree(root.right, subRoot);
    }

    /**
     * Helper function to check if two trees are structurally and valuably identical.
     *
     * @param p The root of the first tree.
     * @param q The root of the second tree.
     * @return True if the two trees are identical, false otherwise.
     */
    private boolean isSameTree(TreeNode p, TreeNode q) {
        // If both nodes are null, they are identical.
        if (p == null && q == null) {
            return true;
        }
        // If one node is null, or their values differ, they are not identical.
        if (p == null || q == null || p.val != q.val) {
            return false;
        }

        // Recursively check the left and right subtrees.
        return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    }
}
